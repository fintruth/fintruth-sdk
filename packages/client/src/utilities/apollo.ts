import { ErrorLink } from 'apollo-link-error'
import { InMemoryCache, defaultDataIdFromObject } from 'apollo-cache-inmemory'

export const createErrorLink = () =>
  new ErrorLink(({ graphQLErrors, networkError }) => {
    if (graphQLErrors) {
      graphQLErrors.map(({ locations, message, path }) =>
        console.warn(
          `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
        )
      )
    }

    if (networkError) {
      console.warn(`[Network error]: ${networkError}`)
    }
  })

export const createInMemoryCache = () =>
  new InMemoryCache({
    dataIdFromObject: obj => {
      switch (obj.__typename) {
        default:
          return defaultDataIdFromObject(obj)
      }
    },
  })
