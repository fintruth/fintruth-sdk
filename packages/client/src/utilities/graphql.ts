import { ProfileInput, RegisterInput } from '@fintruth-sdk/shared'
import { pickAll } from 'ramda'

export const toProfileInput: <T extends ProfileInput>(
  input: T
) => ProfileInput = pickAll(['familyName', 'givenName'])

export const toRegisterInput = <T extends RegisterInput>({
  profile,
  ...input
}: T): RegisterInput => ({
  profile: toProfileInput(profile),
  ...pickAll(['email', 'password'], input),
})
