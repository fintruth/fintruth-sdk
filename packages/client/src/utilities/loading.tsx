import React from 'react'

import Loading from 'components/loading'

interface Options {
  delay?: number
}

export const renderLoadingIf = (
  isLoading: boolean,
  componentFactory: <Props = {}>() => React.ReactElement<Props>,
  options: Options = {}
) => (isLoading ? <Loading {...options} /> : componentFactory())

export const renderLoadingUnless = (
  isLoading: boolean,
  componentFactory: <Props = {}>() => React.ReactElement<Props>,
  options?: Options
) => renderLoadingIf(!isLoading, componentFactory, options)
