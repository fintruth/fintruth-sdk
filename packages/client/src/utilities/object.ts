import { curry, type } from 'ramda'

const isObjectLike = (value: any) =>
  Array.isArray(value) || type(value) === 'Object'

export const omitDeep = curry((keys: string[], obj: {} | any[]) => {
  if (!isObjectLike(obj)) {
    return obj
  }

  if (Array.isArray(obj)) {
    const result: any[] = obj.map(value =>
      isObjectLike(value) ? omitDeep(keys, value) : value
    )

    return result
  }

  const result: { [key: string]: any } = {}

  for (const [key, value] of Object.entries(obj)) {
    if (!keys.includes(key)) {
      result[key] = isObjectLike(value) ? omitDeep(keys, value) : value
    }
  }

  return result
})
