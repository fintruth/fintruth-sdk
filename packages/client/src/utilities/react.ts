import React from 'react'

export const setRef = <T>(
  ref: React.Ref<T> | React.MutableRefObject<T | undefined>,
  instance: T | null
) => {
  if (typeof ref === 'function') {
    ref(instance)
  } else if (ref !== null && instance !== null) {
    // @ts-ignore readonly-property
    ref.current = instance // eslint-disable-line no-param-reassign
  }
}
