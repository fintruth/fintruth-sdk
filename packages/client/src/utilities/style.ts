import {
  Body, // eslint-disable-line import/named
  Code, // eslint-disable-line import/named
  DefaultTheme, // eslint-disable-line import/named
  Hr, // eslint-disable-line import/named
  Pre, // eslint-disable-line import/named
  Html, // eslint-disable-line import/named
  Strong, // eslint-disable-line import/named
  Variables, // eslint-disable-line import/named
} from 'styled-components'
import { em, hsl, hsla, readableColor, rem } from 'polished'

export interface PartialDefaultTheme extends Partial<Variables> {
  body?: Partial<Body>
  code?: Partial<Code>
  hr?: Partial<Hr>
  html?: Partial<Html>
  pre?: Partial<Pre>
  strong?: Partial<Strong>
}

export const createTheme = ({
  grayDarker = hsl(0, 0, 0.21),
  grayDark = hsl(0, 0, 0.29),
  gray = hsl(0, 0, 0.48),
  grayLight = hsl(0, 0, 0.71),
  grayLighter = hsl(0, 0, 0.86),

  whiteTer = hsl(0, 0, 0.96),
  white = hsl(0, 0, 1),

  orange = hsl(14, 1, 0.53),
  yellow = hsl(48, 1, 0.67),
  green = hsl(141, 0.71, 0.48),
  turquoise = hsl(171, 1, 0.41),
  cyan = hsl(204, 0.86, 0.53),
  blue = hsl(217, 0.71, 0.53),
  purple = hsl(271, 1, 0.71),
  red = hsl(348, 1, 0.61),

  backgroundColor = whiteTer,

  primary = turquoise,
  info = cyan,
  success = green,
  warning = yellow,
  danger = red,

  darkContrast = white,
  lightContrast = hsla(0, 0, 0, 0.7),

  blueContrast = readableColor(blue, lightContrast, darkContrast),

  textColor = grayDark,
  textStrongColor = grayDarker,

  gap = 64,

  fontFamilyMonospace = 'monospace',
  fontFamilySansSerif = 'BlinkMacSystemFont, -apple-system, "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue", "Helvetica", "Arial", sans-serif',

  fontFamilyPrimary = fontFamilySansSerif,
  fontFamilyCode = fontFamilyMonospace,

  fontSize4 = rem(24),
  fontSize5 = rem(20),
  fontSize6 = rem(16),
  fontSize7 = rem(12),

  textRendering = 'optimizeLegibility',

  body = {},
  code = {},
  hr = {},
  html = {},
  pre = {},
  strong = {},

  ...rest
}: PartialDefaultTheme = {}): DefaultTheme => ({
  // General Colors
  black: hsl(0, 0, 0.04),
  blackBis: hsl(0, 0, 0.07),
  blackTer: hsl(0, 0, 0.14),

  grayDarker,
  grayDark,
  gray,
  grayLight,
  grayLighter,

  whiteTer,
  whiteBis: hsl(0, 0, 0.98),
  white,

  orange,
  yellow,
  green,
  turquoise,
  cyan,
  blue,
  purple,
  red,

  backgroundColor,

  borderColor: grayLighter,
  borderHoverColor: grayLight,

  // Main Colors
  primary,
  info,
  success,
  warning,
  danger,

  // Contrast Colors
  darkContrast,
  lightContrast,

  orangeContrast: readableColor(orange, lightContrast, darkContrast),
  yellowContrast: readableColor(yellow, lightContrast, darkContrast),
  greenContrast: readableColor(green, lightContrast, darkContrast),
  turquoiseContrast: readableColor(turquoise, lightContrast, darkContrast),
  cyanContrast: readableColor(cyan, lightContrast, darkContrast),
  blueContrast,
  purpleContrast: readableColor(purple, lightContrast, darkContrast),
  redContrast: readableColor(red, lightContrast, darkContrast),

  primaryContrast: readableColor(primary, lightContrast, darkContrast),
  infoContrast: readableColor(info, lightContrast, darkContrast),
  successContrast: readableColor(success, lightContrast, darkContrast),
  warningContrast: readableColor(warning, lightContrast, darkContrast),
  dangerContrast: readableColor(danger, lightContrast, darkContrast),

  // Link Colors
  linkColor: blue,
  linkColorContrast: blueContrast,
  linkVisitedColor: purple,

  linkHoverColor: grayDarker,
  linkHoverBorderColor: grayLight,

  linkFocusColor: grayDarker,
  linkFocusBorderColor: blue,

  linkActiveColor: grayDarker,
  linkActiveBorderColor: grayDark,

  // Text Colors
  textColor,
  textColorContrast: readableColor(textColor, lightContrast, darkContrast),
  textLightColor: gray,
  textStrongColor,
  textSelectionColor: hsl(213, 0.92, 0.85),

  // Responsiveness
  gap,

  viewportSmall: 769,
  viewportMedium: 960 + 2 * gap,
  viewportLarge: 1152 + 2 * gap,
  viewportExtraLarge: 1344 + 2 * gap,

  // Typography
  fontFamilyMonospace,
  fontFamilySansSerif,

  fontFamilyPrimary,
  fontFamilySecondary: fontFamilySansSerif,
  fontFamilyCode,

  fontSize1: rem(48),
  fontSize2: rem(40),
  fontSize3: rem(32),
  fontSize4,
  fontSize5,
  fontSize6,
  fontSize7,

  fontSizeSmall: fontSize7,
  fontSizeNormal: fontSize6,
  fontSizeMedium: fontSize5,
  fontSizeLarge: fontSize4,

  textRendering,

  body: {
    // Colors
    color: textColor,

    // Typography
    fontFamily: fontFamilyPrimary,
    fontWeight: 400,
    lineHeight: 1.4,

    ...body,
  },

  code: {
    // Box
    padding: `${em(4)} ${em(8)}`,

    // Colors
    backgroundColor,
    color: red,

    // Typography
    fontFamily: fontFamilyCode,
    fontSize: `${em(14)}`,
    fontWeight: 400,

    ...code,
  },

  hr: {
    // Box
    height: '2px',
    margin: `${rem(24)} 0`,

    // Colors
    backgroundColor,

    ...hr,
  },

  html: {
    // Colors
    backgroundColor: white,

    // Typography
    fontSize: `${em(16)}`,
    textRendering,

    // Miscellaneous
    overflowX: 'hidden',
    overflowY: 'scroll',

    ...html,
  },

  pre: {
    // Colors
    backgroundColor,
    color: textColor,

    ...pre,
  },

  strong: {
    // Colors
    color: textStrongColor,

    // Typography
    fontWeight: 700,

    ...strong,
  },

  ...rest,
})
