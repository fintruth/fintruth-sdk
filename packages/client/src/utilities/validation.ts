import { StringSchema, string } from '@fintruth-sdk/validation'

import { Type } from 'components/input'

interface InputContext {
  isRequired: boolean
  type?: Type
}

interface SelectContext {
  isRequired: boolean
}

export const validateInput = (value: string, context: InputContext) =>
  string()
    .when(
      ['$isRequired', '$type'],
      (isRequired: boolean, type: Type | undefined, schema: StringSchema) => {
        if (isRequired) {
          schema = schema.required('This is a required field') // eslint-disable-line no-param-reassign
        }

        if (type === 'email') {
          // @ts-ignore missing-properties
          schema = schema.email('Enter a valid email address') // eslint-disable-line no-param-reassign
        } else if (type === 'password') {
          schema = schema.password(2) // eslint-disable-line no-param-reassign
        } else if (type === 'tel') {
          schema = schema.phone('Enter a valid phone number') // eslint-disable-line no-param-reassign
        }

        return schema
      }
    )
    .validate(value, { context })
    .then(() => '')
    .catch(error => error.message)

export const validateSelect = (value: string, context: SelectContext) =>
  string()
    .when('$isRequired', (isRequired: boolean, schema: StringSchema) =>
      isRequired ? schema.required('This is a required field') : schema
    )
    .validate(value, { context })
    .then(() => '')
    .catch(error => error.message)
