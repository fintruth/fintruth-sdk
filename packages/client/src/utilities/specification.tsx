import {
  LocationProvider,
  LocationProviderRenderFn,
  createHistory,
  createMemorySource,
} from '@reach/router'
import { render } from '@testing-library/react'
import React from 'react'

interface RouterOptions {
  initialPath?: string
}

export const renderWithRouter = (
  node: React.ReactNode | LocationProviderRenderFn,
  { initialPath = '/' }: RouterOptions = {}
) => {
  const history = createHistory(createMemorySource(initialPath))

  return {
    ...render(<LocationProvider history={history}>{node}</LocationProvider>),
    history,
  }
}
