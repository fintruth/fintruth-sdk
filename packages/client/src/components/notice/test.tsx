import React from 'react'
import { ThemeProvider } from 'styled-components'
import { render } from 'react-testing-library'

import { createTheme } from 'utilities/style'
import Notice from '.'

test('should render the default variant correctly', () => {
  const { container } = render(
    <ThemeProvider theme={createTheme()}>
      <Notice>child</Notice>
    </ThemeProvider>
  )

  expect(container.firstChild).toMatchSnapshot()
})

test('should render the danger variant correctly', () => {
  const { container } = render(
    <ThemeProvider theme={createTheme()}>
      <Notice variant="danger">child</Notice>
    </ThemeProvider>
  )

  expect(container.firstChild).toMatchSnapshot()
})

test('should render the success variant correctly', () => {
  const { container } = render(
    <ThemeProvider theme={createTheme()}>
      <Notice variant="success">child</Notice>
    </ThemeProvider>
  )

  expect(container.firstChild).toMatchSnapshot()
})
