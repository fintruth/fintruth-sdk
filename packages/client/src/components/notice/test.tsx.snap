// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the danger variant correctly 1`] = `
.c0 {
  color: #ff3860;
  display: block;
  font-size: 0.75rem;
  margin-top: 0.25rem;
}

<p
  class="c0"
>
  child
</p>
`;

exports[`should render the default variant correctly 1`] = `
.c0 {
  color: #4a4a4a;
  display: block;
  font-size: 0.75rem;
  margin-top: 0.25rem;
}

<p
  class="c0"
>
  child
</p>
`;

exports[`should render the success variant correctly 1`] = `
.c0 {
  color: #23d160;
  display: block;
  font-size: 0.75rem;
  margin-top: 0.25rem;
}

<p
  class="c0"
>
  child
</p>
`;
